### DISCUSSION TITLE
How to add humidity and pressure to temperature sensors?

### DISCUSSION THREAD

**Discussion by pbawesome** [freakent/dbus-mqtt-devices#136] [Q&A]
I followed the node red example for temperature sensor. How can I also add pressure and humidity to the sensor I already created using the example in node red?

please hold my hand a little bit as I’m kinda new to this.

**Comment by freakent** [comment]
Is it the dbus-mqtt-devices flow sample that sends Rasperry Pi's CPU temperature and free space that you are using?  Send us a screen shot, but I'll assume that it is.

When you register a temperature device using the driver is defines 3 value attributes on the dbus which Victron's Venus OS is expecting; Temperature, Pressure and Humidity (note the capilatisation, it is important).  You can find these attributes defined in the ModBus TCP register spreadhseet that Victron provides and also in the [services.yml](https://github.com/freakent/dbus-mqtt-devices/blob/main/services.yml) file in this driver. Victron designed their MQTT interface to mirror the dbus attributes. You have to publish a message to the relavent topic for each attribute you want to populate.  

In the sample, look at the "Prepare Msgs" change node that receives temperature data. It sets the "value" attribute of msg.payload to the temperature value and the msg.topic to "W/{portalId}/temperature/{device id}/Temperature". "W" are write topics, if you don't yet know difference between Victron's W, R and N topics then check their [MQTT documentaion](https://github.com/victronenergy/dbus-flashmq). If you want to send pressure you must also send a msg.payload with a value attribute to "W/{portalId}/temperature/{device id}/Pressure" and similarly for Humidity. To send Temperature, Humidity and Pressure you would have publish 3 messages.  It's a very good idea to use MQTT Explorer (but aldo read the section on keepalive in Victron's documentaion) and dbus-spy so that you can see what's happening.

What I described above is the standard Victron MQTT api. Personally I find publishing multiple messages, one for each attribute, a little cumbersome so to simplify things I recently added a simple proxy to the driver. The proxy takes the same information but in one json payload and splits it up into the multiple messages that Victron's api require. For example:
To send Temperature, Humidity and Pressure using the proxy you would publish the following msg.payload 
```
{
   "topicPath": "W/<portalid>>/temperature/<deviceid>", # Note the W for a write topic
   "values": {
       "Temperature" : 23,
       "Pressure": 900,
       "Humidity": 56
    }
}
```
to topic `device/<clientId>/Proxy`
The proxy then breaks that into 3 messages and publsihes each on your behalf. Use of the proxy is entirely optional and is personal preference. There's more info in the README, including how the driver provides you with the topic path string.
